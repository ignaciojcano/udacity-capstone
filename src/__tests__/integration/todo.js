const supertest = require("supertest");
const faker = require("faker");
const Todo = require("../../models/todo");
const server = require("../../server");

const request = supertest(server);

const createTodo = async (id, complete = false) =>
  Todo.create({
    _id: id,
    text: faker.lorem.sentence(),
    complete
  });

const removeDateFields = todo => {
  delete todo.createdAt;
  delete todo.updatedAt;
  return todo;
};

describe("TODOS", () => {
  beforeEach(() => {
    faker.seed(156879);
  });

  afterEach(async () => {
    await Todo.collection.drop();
  });

  it("should fetch all the todos", async () => {
    await createTodo("56cb91bdc3464f14678934cd");
    await createTodo("56cb91bdc3464f14678934cb", true);
    await createTodo("56cb91bdc3464f14678934cc");
    const response = await request.get("/todos");
    expect(response.status).toBe(200);
    expect(response.body.map(removeDateFields)).toMatchSnapshot();
  });

  it("should create a todo", async () => {
    const text = "A todo description";
    const response = await request.post("/todos").send({ text });
    expect(response.status).toBe(200);
    const { body } = response;
    expect(response.body._id).toHaveLength(24); // valid object id
    delete body._id; // because this changes everytime
    expect(removeDateFields(body)).toMatchSnapshot();
  });

  it("should fetch a todo by id", async () => {
    const id = "56cb91bdc3464f14678934ca";
    await createTodo(id);
    const response = await request.get(`/todos/${id}`);
    expect(response.status).toBe(200);
    expect(removeDateFields(response.body)).toMatchSnapshot();
  });

  it("should update a todo by id", async () => {
    const id = "56cb91bdc3464f14678934ca";
    await createTodo(id);
    const text = "An updated description";
    const response = await request
      .put(`/todos/${id}`)
      .send({ text, complete: true });
    expect(response.status).toBe(200);
    expect(removeDateFields(response.body)).toMatchSnapshot();
  });
});
