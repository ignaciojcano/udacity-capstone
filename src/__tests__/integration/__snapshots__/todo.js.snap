// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TODOS should create a todo 1`] = `
Object {
  "__v": 0,
  "text": "A todo description",
}
`;

exports[`TODOS should fetch a todo by id 1`] = `
Object {
  "__v": 0,
  "_id": "56cb91bdc3464f14678934ca",
  "complete": false,
  "text": "Veniam et esse autem et.",
}
`;

exports[`TODOS should fetch all the todos 1`] = `
Array [
  Object {
    "__v": 0,
    "_id": "56cb91bdc3464f14678934ca",
    "complete": false,
    "text": "Veniam et esse autem et.",
  },
  Object {
    "__v": 0,
    "_id": "56cb91bdc3464f14678934cb",
    "complete": true,
    "text": "Eligendi non mollitia.",
  },
  Object {
    "__v": 0,
    "_id": "56cb91bdc3464f14678934cc",
    "complete": false,
    "text": "Sed quisquam neque quibusdam quo ullam labore magnam ullam optio.",
  },
]
`;

exports[`TODOS should update a todo by id 1`] = `
Object {
  "__v": 0,
  "_id": "56cb91bdc3464f14678934ca",
  "complete": true,
  "text": "An updated description",
}
`;
